### DISCUSSION TITLE
Laravel

### DISCUSSION THREAD

**Discussion by axelrindle** [Leantime/leantime#1060] [Ideas]
Laravel 9 has been around for quite a while now and I'd like to discuss the idea of migrating or rebuilding Leantime on top of Laravel at some point in the future. 

- **Framework**: As a framework, Laravel does nothing on its own. It gives you building blocks to build applications. These building blocks are powerful and require few configuration to work. There are very many of these building blocks for almost everything: _Authentication_, _Authorization_, _Session_, _Database_, _Storage_, _Mailing_, _Events_, Scheduling, etc.
- **Community**: Being probably *the* most popular Php framework out there, Laravel is accompanied by a huge community and thousands of custom extensions and packages for it. Here are some packages I like:
  - https://github.com/spatie/laravel-backup
  - https://github.com/spatie/laravel-activitylog
  - https://github.com/the-control-group/voyager
  - https://github.com/filamentphp/filament
  - _These are awesome too_: https://github.com/chiraggude/awesome-laravel
  - _There is more_: https://github.com/topics/laravel
- **Ecosystem**: Laravel features a large ecosystem of extensions and official packages for complex workloads. Here are the ones I like the most:
  - _Echo_: Realtime communication using WebSockets.
  - _Horizon_: Ready-to-use UI for queue monitoring.
  - _Sanctum_: Application authentication.
  - _Scout_: Full-text search with support for multiple engines.
  - _Socialite_: Social login.

> Laravel gives you robust tools for [dependency injection](https://laravel.com/docs/9.x/container), [unit testing](https://laravel.com/docs/9.x/testing), [queues](https://laravel.com/docs/9.x/queues), [real-time events](https://laravel.com/docs/9.x/broadcasting), and more. Laravel is fine-tuned for building professional web applications and ready to handle enterprise workloads.
>
> -- From [laravel.com](https://laravel.com/docs/9.x/installation#why-laravel)

Here are some pretty big projects built using Laravel:

- https://github.com/Attendize/Attendize
- https://github.com/monicahq/monica
- https://github.com/koel/koel
- https://github.com/flarum/flarum
- https://github.com/CachetHQ/Cachet
- https://github.com/invoiceninja/invoiceninja
- https://github.com/bagisto/bagisto

I am aware of the fact that such a migration/rebuild won't be an easy task as Leantime is quite complex already and uses an entirely custom structure. But thinking of mid- and long-term maintainability, I'm firmly convinced that using production-proven software will make everyone's life easier.

This is not meant to be any kind of advertisement for Laravel. I tried to point out the key facts why Laravel is a good thing in my eyes. I already use(d) Laravel in several projects and would be happy to share some of my experience.

Don't get me wrong: Leantime is awesome! I use it for a small "side-business" I'm currently building and want to be a part of Leantime's future. I'd like to see Leantime grow further in the future and this might be a huge improvement on the way.

**Comment by marcelfolaron** [comment]
Hi, 

thanks I appreciate the post. While I agree that Laravel has a large following and is at the forefront of PHP framework standardization I don't think it is worth the effort to migrate the entire Leantime platform to Laravel.

I don't really see the the return on our investment and I'm sure a port would take more than 1 year to finish. 
What are the actual benefits that would warrant this type of effort and holding off on any feature development? 

The Leantime framework has gone through multiple iterations and I strongly believe that we have a solid microframework base that allows us to run Leantime faster than we'd be able to on Laravel.

Happy to discuss further and get more thoughts though.

**Comment by gloriafolaron** [comment]
Just wanted to chime in on this as I wasn't familiar with Laravel and it looks like there's a lot of value in migrating to something like that in terms of long term scalability and I appreciate you taking the time to detail it out and sharing that value.

Some questions that come up:

What kind of time frame do you think this would be most useful in as a user of the system in terms of execution?
When you say mid and long term maintainability, could you touch on where you see this helping the most? (User contributions, user modifications, general updating, etc)

To Marcel's point, migrating right now would be a layer of complexity that we couldn't handle internally here but there's definitely value in the longer term discussion around it and also in conversations around making Leantime's framework more accessible (as I think I hear that's what the goal is?). Would love to hear your thoughts.

Thanks for starting the conversation!
